### DISCUSSION TITLE
Introducing AnimaTrip, an application for low-latency webcam video

### DISCUSSION THREAD

**Discussion by xnakos** [jamulussoftware/jamulus#3154] [Ideas]
Hi, Jamulus users!

On https://github.com/xnakos/animatrip I have pushed an extended experiment of mine for the last couple of months, aiming to provide low-latency webcam video for multiple users in a simple way as a client-server application. I was greatly influenced by JackTrip, hence the name to pay... tripute! Of course I had also come across and experimented with Jamulus, the brilliance of which I had admired from the start. My target was something complementary to low-latency audio for online performance, from the webcam video aspect. For that I used the fantastic GStreamer library and tools. AnimaTrip functionality is currently limited to Linux-only. You also have to build it on your own, first installing the `gstreamer` and `gio` dependencies. Maybe the application proves useful to some of you or just raises questions about low-latency video. Discussions, issues, etc., are welcome.

Thanks and keep jamming!
Harry

**Comment by mcfnord** [comment]
I've been fishing around for a video option to add to any Jamulus server. I might recompile Jitsi with audio off. I do need a client-server design that embeds in a browser, so the custom client build wouldn't work. But there is a need for a lightweight video-only teleconference solution.

**Reply by xnakos** [reply]
Thank you for the comment. I have not used Jitsi and I would be interested in what latencies are achieved. In general I am missing what latencies are required and for what purposes. Like, for instance, less that 10ms are required for effortlessly playing while listening to yourself, less than 25-30ms are required for effortlessly playing with others, etc. As for the browser-embeddable client, that makes a lot of sense.

**Comment by ann0see** [comment]
Sounds interesting. If the latency is good enough to follow a conductor that would change quite a lot.

**Reply by xnakos** [reply]
Thank you for the insight. I did not have a conductor in mind and now wonder what an acceptable latency for that would be. I mostly had a drummer in mind. :)

**Comment by dcorson-ticino-com** [comment]
"Low Latency" is a very spongy concept depending on who you are talking too.
Do you have any measured times that you could share with us?
What image update rate are you working at?

**Reply by xnakos** [reply]
You are totally right about the low-latency term. The reason for the experiment is something to visually accompany JackTrip or Jamulus, so that, ideally, even visual synchronization between jammers could be possible. So, just general sub-second latency would not be enough. There must be research about such synchronization, which I have not investigated. On the other hand, I have used what is available by GStreamer 1.20, making communication as quick and rudimentary as possible in code. There is definitely some room for improvement in certain areas. I have mostly used 30FPS for the camera and 60Hz for the screen refresh rate. These are the lowest values possible and can be increased. Using these camera and screen numbers, locally launching the server and a client, I have taken measurements of 114, 148, 170, 199, 135, 168, 148, and 178 milliseconds of latency. The measurement method was a bit unorthodox, yet somewhat real-life: In my laptop I started the millisecond-accurate `stopwatch` and put it under the camera. I then started screen video capturing, took a small mirror and sent the stopwatch image to the camera. Visiting video frames, in each frame I subtracted the client stopwatch value from the screen stopwatch value to get an estimate of the latency, resulting in the values above. An example frame is this:

![g-148ms](https://github.com/jamulussoftware/jamulus/assets/1213128/94f05fa6-ac4b-4ec4-b864-d9189e1add0d)

Another frame is this:

![h-178ms](https://github.com/jamulussoftware/jamulus/assets/1213128/bb071199-f956-40af-a496-7aa01e101c34)

Note the same value in the client stopwatch value (I chose such consecutive frames on purpose), because of the reduced camera FPS. Camera FPS and screen refresh rate define the error margin of the measurement.

I am currently on vacation, breaking a bit of vacation protocol right now :) . After I get back, I will investigate more.

**Reply by dcorson-ticino-com** [reply]
It is difficult to say what is really required as all the alternatives now (that I know at least) are way too slow, so can be rejected out of hand.  We have no experience to go by.  Zoom or Jitzi, for example, could never be used for more than the communal feeling of playing together.
I think that one of the reasons Jamulus works so well even with relatively high latencies (by high latency I mean total delay >90ms) is that we don't see the other performers and base our own playing on what we hear coming back from the server mix.
To keep things sane I think it would be necessary to have a visual delay similar to the audio delay.  If we assume the video server is on the same computer as the audio server we will have similar communications delay.  That would mean the total round trip video processing and buffering delay should be similar to the audio processing and buffering delay ~10 to 15ms in my case.
That seems to be a lot to ask.  
But even just using video for the atmospheric fun, not trying to base our playing on it, has value.
So keep up the good work.
(It may, however, not be possible to make the choir directors happy)

**Reply by mcfnord** [reply]
I agree that conductor-level sync between audio and video would be a breakthrough, and I agree that missing this mark and just reaching "atmospheric fun" would still be very valuable.

I'd like to configure Jitsi for our particular needs: No sound, lowest latency, probably sacrificing pixel count and frame consistency. I believe Jitsi is also UDP and client-server, which suggests its performance could be similar to a Jamulus server. But Jitsi is a bigger server than Jamulus, with higher costs, so it won't be widely deployed for lowest latency the way Jamulus server is. (A no-sound recompile could change that.) Ideally, a video-only server could deploy with each Jamulus server, because the latency of the two channels would be very similar. Maybe video could even throttle itself when the Jamulus server reports performance degradation.

Some commercial competitors offer video, and most say it's a "nice-to-have" but a basic integration that prioritizes audio (of course) would be a breakthrough for us.

**Reply by ann0see** [reply]
@mcfnord do you know https://jamulus.io/kb/2021/04/18/Jamulus-Video.html

**Reply by chrisrimple** [reply]
It's not good for conducting a group or other real-time needs, but the [WorldJam](http://worldjam.vip) uses OBS to livestream Jamulus + Zoom to YouTube, Twitch, etc. with the audio stream delayed about 250ms to sync with the video stream.

**Reply by ann0see** [reply]
@chrisrimple it's probably worldjam.vip

But I think jitsi is sometimes ok ish if the conductor just gives dynamic or non time critical signs.

**Reply by chrisrimple** [reply]
Oops, yes, fixed.  I was thinking of the Jamulus website, which ends in .io  :)

**Comment by gene96817** [comment]
Thank you for all your work. This past year I only have had time to lurk.

The speed of sound is about 3 ms per meter (1 ms per foot).
Consider the distance many musicians are from a conductor. Many of us are easily 20-40ms away from the conductor.

There is plenty of margin for the video to be useful.

Gene
----------------------------
Eugene Chang
Honolulu, Hawaii
***@***.***
